src/SourceCode.Chasm/TreeNodeMap.cs

```diff
@@ -175,6 +179,53 @@ public TreeNodeMap Merge(TreeNode node)
             return new TreeNodeMap(array);
         }
 
+        public TreeNodeMap Delete(Func<string, bool> predicate)
+        {
+            if (predicate == null) throw new ArgumentNullException(nameof(predicate));
+            if (_nodes.Length == 0) return this;
+
+            var copy = new TreeNode[_nodes.Length - 1];
+            var span = _nodes.Span;
+            var j = 0;
+            for (var i = 0; i < _nodes.Length; i++)
+            {
+                if (!predicate(span[i].Name))
+                    copy[j++] = span[i];
+            }
+
+            if (j == _nodes.Length)
+                return this;
+
+            return new TreeNodeMap(new ReadOnlyMemory<TreeNode>(copy, 0, j));
+        }
+
+        public TreeNodeMap Delete(string key)
+        {
+            if (_nodes.Length == 0) return this;
+
+            var copy = new TreeNode[_nodes.Length - 1];
+            var span = _nodes.Span;
+            var found = false;
+            for (var i = 0; i < _nodes.Length; i++)
+            {
+                if (found)
+                {
+                    copy[i - 1] = span[i];
+                }
+                else
+                {
+                    if (i < _nodes.Length - 1)
+                        copy[i] = span[i];
+                    found = string.Equals(span[i].Name, key, StringComparison.Ordinal);
```

Comment:
`StringComparer.Ordinal.Equals` is faster - your code just calls that eventually anyway (after lots of branching)